src/libostree/ostree-sysroot-deploy.c

```diff
@@ -2824,34 +2837,250 @@ get_var_dfd (OstreeSysroot      *self,
+            continue;
+
+          const char *karg_name = dent->d_name;
+          if (g_tree_lookup (kargs_configs, karg_name))
+            continue;
+
+          g_autofree char *karg_contents = NULL;
+          if (!(karg_contents = glnx_file_get_contents_utf8_at (kargs_dfd,
+                                                                karg_name, NULL,
+                                                                cancellable,
+                                                                error)))
+            return FALSE;
 
+          g_tree_insert (kargs_configs, g_strdup (karg_name),
+                         g_steal_pointer (&karg_contents));
+        }
+    }
+
+  return TRUE;
+}
+
+static gboolean
+add_kargs_from_commit (OstreeSysroot *sysroot,
+                       GTree         *kargs_configs,
+                       int            deployment_dfd,
+                       const char    *dirpath,
+                       const char    *revision,
+                       GCancellable  *cancellable,
+                       GError       **error)
+{
+  struct stat stbuf;
+  if (!glnx_fstatat_allow_noent (deployment_dfd, dirpath, &stbuf, 0, error))
+    return FALSE;
+  const gboolean dir_exists = (errno == 0);
+
+  if (dir_exists)
+    {
+      g_autoptr(GFile) karg_file = get_file_from_repo (ostree_sysroot_repo (sysroot),
+                                                       revision,
+                                                       dirpath,
+                                                       cancellable,
+                                                       error);
+      g_autoptr(GFileEnumerator) dir_enum = NULL;
+      g_autoptr(GFile) child = NULL;
+      g_autoptr(GFileInfo) child_info = NULL;
+      g_autoptr(GError) temp_error = NULL;
+      dir_enum = g_file_enumerate_children (karg_file,
+                                            OSTREE_GIO_FAST_QUERYINFO,
+                                            G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                                            cancellable, error);
+      if (!dir_enum)
+        return FALSE;
+
+      while ((child_info = g_file_enumerator_next_file (dir_enum, NULL,
+                                                        &temp_error)) != NULL)
+        {
+          g_clear_object (&child);
+          const char *karg_name = g_file_info_get_name (child_info);
+          child = g_file_get_child (karg_file, karg_name);
+
+          g_autofree char *karg_contents = NULL;
+          if (!g_file_load_contents (child, cancellable, &karg_contents, NULL,
+                                     NULL, &temp_error))
+            {
+              g_propagate_error (error, g_steal_pointer (&temp_error));
+              return FALSE;
+            }
+          if (karg_contents && !g_tree_lookup (kargs_configs, karg_name))
+            g_tree_insert (kargs_configs, g_strdup (karg_name),
+                           g_steal_pointer (&karg_contents));
```

Comment:
The way this currently works, if there is a host configured karg in `/etc/ostree/kargs.d`, and the vendor adds the *same* karg at some point in the future during an update but with a different name for the karg file, then there will be a duplicates in the BLS entry. 
Additionally, it is possible that there are multiple host configured kargs all in the _same_ karg file, and when the vendor adds a subset of these kargs at some point during an update, we would have duplicates.

To deal with these two potential problems, perhaps we can make it so that each karg file must only have one karg and the karg KEY must be in the name of the karg file. However, then we would need to have some extra logic to ignore the number preceding the karg file or add the ordering info to the karg file.